identity-service/src/solana-client.js

```diff
@@ -191,19 +190,125 @@ async function createAndVerifyMessage (
   })
 
   let feePayerAccount = getFeePayerKeypair(false)
+  let signature = await sendAndSignTransaction(connection, transaction, feePayerAccount)
 
-  let signature = await solanaWeb3.sendAndConfirmTransaction(
-    connection,
-    transaction,
-    [feePayerAccount],
+  return signature
+}
+
+const getUnixTs = () => {
+  return new Date().getTime() / 1000;
+};
+
+// Adapted from mango send function
+// https://github.com/blockworks-foundation/mango-ui/blob/b6abfc6c13b71fc17ebbe766f50b8215fa1ec54f/src/utils/send.tsx#L785
+async function sendAndSignTransaction (connection, transaction, signers) {
+  let recentBlockHash = (await connection.getRecentBlockhash('confirmed')).blockhash
+  transaction.recentBlockhash = recentBlockHash
+
+  transaction.sign(signers)
+  let rawTransaction = transaction.serialize()
+  const startTime = getUnixTs();
+  const txid = await connection.sendRawTransaction(
+    rawTransaction,
     {
-      skipPreflight: false,
-      commitment: config.get('solanaTxCommitmentLevel'),
-      preflightCommitment: config.get('solanaTxCommitmentLevel')
+      skipPreflight: true,
+    },
+  );
+
+  // Anonymous function to retry sending until confirmation
+  (async () => {
+    while (!done && getUnixTs() - startTime < timeout) {
+      connection.sendRawTransaction(rawTransaction, {
+        skipPreflight: true,
+      });
+      await sleep(300);
     }
-  )
+  })();
 
-  return signature
+  try {
+    awaitTransactionSignatureConfirmation(txid, 1000, connection)
+  } catch(e) {
+    throw new Error(e)
+  } finally {
+    done = true
+  }
+  return txid
+}
+
+async function awaitTransactionSignatureConfirmation(
+  txid,
+  timeout,
+  connection,
+) {
+  let done = false;
+  const result = await new Promise((resolve, reject) => {
+    (async () => {
+      setTimeout(() => {
+        if (done) {
+          return;
+        }
+        done = true;
+        console.log('Timed out for txid', txid);
+        reject({ timeout: true });
+      }, timeout);
+      try {
+        connection.onSignature(
+          txid,
+          (result) => {
+            console.log('WS confirmed', txid, result);
+            done = true;
+            if (result.err) {
+              reject(result.err);
+            } else {
+              resolve(result);
+            }
+          },
+          connection.commitment,
+        );
+        console.log('Set up WS connection', txid);
```

Comment:
I'm assuming that the onSignature is an event listener callback and is irrelevant to setting up a ws connection?